File: bench/stream_client.ts
#!/usr/bin/env -S npx tsx
/**
 * SSE streaming benchmark — TypeScript
 * Measures: connect → fetch stream → read lines → JSON.parse → extract
 */

const PORT = parseInt(process.env.PORT || '3000', 10);
const EVENTS = parseInt(process.env.STREAM_EVENTS || '10000', 10);

interface StreamResult {
  lang: string;
  bench: string;
  events: number;
  extracted: number;
  total_ms: number;
  events_per_sec: number;
}

async function benchStream(provider: string): Promise<StreamResult> {
  const pathMap: Record<string, string> = {
    openai: '/openai/stream',
    anthropic: '/anthropic/stream',
    ollama: '/ollama/stream',
  };

  const url = `http://127.0.0.1:${PORT}${pathMap[provider]}?n=${EVENTS}`;

  const start = performance.now();
  const response = await fetch(url);
  const reader = response.body!.getReader();
  const decoder = new TextDecoder();

  let eventCount = 0;
  let extractCount = 0;
  let buffer = '';

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split('\n');
    buffer = lines.pop()!; // keep incomplete last line

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed.startsWith('data: ')) continue;
      const data = trimmed.slice(6);
      if (data === '[DONE]') continue;

      eventCount++;

      try {
        const parsed = JSON.parse(data);
        if (provider === 'openai') {
          if (parsed.choices?.[0]?.delta?.content != null) extractCount++;
        } else if (provider === 'anthropic') {
          if (parsed.delta?.text != null) extractCount++;
        } else {
          if (parsed.response != null) extractCount++;
        }
      } catch {}
    }
  }

  const elapsed = performance.now() - start;
  return {
    lang: 'typescript',
    bench: `sse_${provider}`,
    events: eventCount,
    extracted: extractCount,
    total_ms: Math.round(elapsed * 100) / 100,
    events_per_sec: Math.round(eventCount / (elapsed / 1000)),
  };
}

async function main() {
  for (const provider of ['openai', 'anthropic', 'ollama']) {
    const result = await benchStream(provider);
    console.log(JSON.stringify(result));
  }
}

main().catch(console.error);

File: bench/stream_client.py
#!/usr/bin/env python3
"""SSE streaming benchmark — Python
Measures: connect → read stream → split lines → json.loads → extract
"""

import json
import os
import time
import urllib.request

PORT = int(os.environ.get('PORT', '3000'))
EVENTS = int(os.environ.get('STREAM_EVENTS', '10000'))


def bench_stream(provider):
    path_map = {
        'openai': '/openai/stream',
        'anthropic': '/anthropic/stream',
        'ollama': '/ollama/stream',
    }

    url = f'http://127.0.0.1:{PORT}{path_map[provider]}?n={EVENTS}'

    start = time.perf_counter_ns()
    req = urllib.request.urlopen(url)

    event_count = 0
    extract_count = 0

    for raw_line in req:
        line = raw_line.decode('utf-8').strip()
        if not line.startswith('data: '):
            continue
        data = line[6:]
        if data == '[DONE]':
            break

        event_count += 1

        try:
            parsed = json.loads(data)
            if provider == 'openai':
                if parsed.get('choices', [{}])[0].get('delta', {}).get('content') is not None:
                    extract_count += 1
            elif provider == 'anthropic':
                if parsed.get('delta', {}).get('text') is not None:
                    extract_count += 1
            else:
                if parsed.get('response') is not None:
                    extract_count += 1
        except json.JSONDecodeError:
            pass

    req.close()
    elapsed_ns = time.perf_counter_ns() - start
    total_ms = elapsed_ns / 1_000_000
    events_per_sec = int(event_count / (elapsed_ns / 1_000_000_000)) if elapsed_ns > 0 else 0

    result = {
        'lang': 'python',
        'bench': f'sse_{provider}',
        'events': event_count,
        'extracted': extract_count,
        'total_ms': round(total_ms, 2),
        'events_per_sec': events_per_sec,
    }
    print(json.dumps(result))


if __name__ == '__main__':
    for provider in ['openai', 'anthropic', 'ollama']:
        bench_stream(provider)
